### PR TITLE
codemod: add transform replaceReactUseRef

### LINKED ISSUE #133: codemod: prettier is a runtime dependency
https://github.com/swordev/suid/blob/6945484938e68c929a4540ea502931445d40ed22/packages/codemod/package.json#L28-L29

```
cd $(mktemp -d)
npm init -y
npm i @suid/codemod
npx @suid/codemod
```

> ```Error [ERR_MODULE_NOT_FOUND]: Cannot find package 'prettier' imported from /tmp/tmp.kXmaXvmPTv/node_modules/@suid/codemod/actions/mui2suid.js```

### LINKED ISSUE #135: move snapshot tests to separate files
currently we store multiple test cases together

```js
  it("transforms a to b", () => {
    expect(
      t(`
        a
      `)
    ).toBe(
      format(`
        b
      `)
    );
  });
  it("transforms c to d", () => {
    expect(
      t(`
        c
      `)
    ).toBe(
      format(`
        d
      `)
    );
  });
```

... for example in `packages/codemod/test/transforms/replaceReactUseEffect.test.ts`

make this more pretty by storing the test cases in separate files
for example in `packages/codemod/test/transforms/replaceReactUseEffect/`

`transforms a to b.src.ts`

```
a
```

`transforms a to b.out.ts`

```
b
```

> make this more pretty

benefit: pretty `git diff`

similar issue https://github.com/BuilderIO/mitosis/issues/856

compromise solutions with `jest-snapshot` ...

```js
expect("a").toMatchSnapshot() // snapshot is stored in __snapshots__/*.snap
expect("a").toMatchInlineSnapshot("b")
```

&rarr; update snapshots with `npx jest -u`

https://jestjs.io/docs/snapshot-testing

### LINKED ISSUE #136: handle default import: import React from "react"
https://github.com/swordev/suid/blob/6945484938e68c929a4540ea502931445d40ed22/packages/codemod/src/navigations/findReactObjects.ts#L16

should be

```ts
    const ns = (
      reactImport.getNamespaceImport() || // import * as React from "react"
      reactImport.getDefaultImport() // import React from "react"
    );
```

### PR DESCRIPTION
**Pull request by milahu**
note: looking back, react2solid should be implemented as an eslint plugin
for example https://github.com/milahu/eslint-plugin-react2solid
why? eslint is the most popular transformer = large ecosystem = reuse existing rules
typescript is slow (except we use swc), typescript should be optional

transform `React.useRef`

```diff
-const divRef = React.useRef<HTMLDivElement>(null)
+var divRef: HTMLDivElement | null = null;
const div = <div ref={divRef}>asdf</div>
-console.log(divRef.current)
+console.log(divRef)
```

not perfect but should work for most cases

also

fix #133 
fix #136 

fix #135 (partial) - use `toMatchInlineSnapshot` to allow `npx jest -u`

### CODE REVIEW

File: packages/codemod/test/transforms/replaceReactUseRef.test.ts [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+      findReactObjects(source, ["useRef"]).map((v) =>
+        replaceReactUseRef(v.node)
+      ),
+  ]);
+
+describe("replaceReactUseRef", () => {
+  it("transforms to let", () => {
+    expect(
+      t(`
+        import React from 'react';
```
Comment (milahu) [inline packages/codemod/test/transforms/replaceReactUseRef.test.ts:19]: depends on #136

File: packages/codemod/test/transforms/replaceReactUseRef.test.ts [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+              <div>noref6 = {noref6}</div>
+              <div>ref7.current = {ref7.current}</div>
+              <div>ref8.current = {ref8.current}</div>
+            </div>
+          )
+        }
+      `)
+    ).toBe(
+      format(`
+        todo
```
Comment (milahu) [inline packages/codemod/test/transforms/replaceReactUseRef.test.ts:93]: example output with all transforms

<details>

```tsx
import { createSignal, createEffect, on } from "solid-js";

let topLevelRef = 0;
let topLevelNoRef1 = 1;
let topLevelRef2 = "2";
let topLevelNoRef3 = 3;
let topLevelRef4 = [4];
console.log(topLevelRef)
console.log(topLevelNoRef1)
console.log(topLevelRef2)
console.log(topLevelNoRef3)
console.log(topLevelRef4)

export default function App(props: any) {
    console.log(props)
    const [someState, setSomeState] = createSignal(0)
    let divRef: HTMLDivElement | null = null;
    let divRefDefaultUndefined: HTMLDivElement | undefined = undefined;
    let divRefDefaultString: HTMLDivElement | string = "asdf";
    let divRefDefaultNumber: HTMLDivElement | number = 123;
    let divRefDefaultTrue: HTMLDivElement | boolean = true;
    let divRefDefaultFalse: HTMLDivElement | boolean = false;
    console.log(divRef) // access in same scope
    setSomeState(1)
    console.log(someState())
    let ref1 = "1";
    let noref2 = "2";
    let ref3: number = 3;
    let noref4 = 4;
    let ref5: number | null = 5;
    let noref6 = 6;
    let ref7: number | null | undefined = 7;
    let ref8 = 8;
    const [counter1, setCounter1] = createSignal(props.init1 || 0)
    const [counter2, setCounter2] = createSignal(props.init2 || 0)
    const increment1 = () => setCounter1(counter1() + 1)
    const increment2 = () => setCounter2(counter2() + 1)
    const colors = ["red", "yellow", "green", "blue"]
    const [colorId, setColorId] = createSignal(0)
    createEffect(on(
        () => [counter1()],
        () => {
            const div = divRef // access in child scope
            if (div) {
                setColorId((colorId() + 1) % 3)
                div.style.color = colors[colorId()]
            }
        }
    ))
    return (
        <div
            ref={divRef}
            style={{
                color: "white",
                background: "black",
                width: "80vw",
                "user-select": "none"
            }}
        >
            <h2>App</h2>
            <div ref={divRefDefaultUndefined}>divRefDefaultUndefined</div>
            <div ref={divRefDefaultString}>divRefDefaultString</div>
            <div ref={divRefDefaultNumber}>divRefDefaultNumber</div>
            <div ref={divRefDefaultTrue}>divRefDefaultNumber</div>
            <div ref={divRefDefaultFalse}>divRefDefaultNumber</div>
            <div onClick={increment1}>counter1 = {counter1()} (click me)</div>
            <div onClick={increment2}>counter2 = {counter2()} (click me)</div>
            <div>ref1.current = {ref1}</div>
            <div>noref2 = {noref2}</div>
            <div>ref3.current = {ref3}</div>
            <div>noref4 = {noref4}</div>
            <div>ref5.current = {ref5}</div>
            <div>noref6 = {noref6}</div>
            <div>ref7.current = {ref7}</div>
            <div>ref8.current = {ref8}</div>
        </div>
    )
}
```

</details>

**Comment by milahu** [comment]
41e39e7a2f7ee7a08fcec7cac189b59ec80ba856 causes regression in

```tsx
import * as React from 'react'

interface DevtoolsPanelOptions extends ContextOptions {
  showCloseButton?: boolean
}

const ReactQueryDevtoolsPanel = (
    React.forwardRef<HTMLDivElement, DevtoolsPanelOptions>(
        function ReactQueryDevtoolsPanel(props, ref) {
            const { showCloseButton } = props
            return <div></div>
        }
    )
)

export function ReactQueryDevtools() {
  return <ReactQueryDevtoolsPanel showCloseButton/>
}
```

trying to replace

```diff
-  return <ReactQueryDevtoolsPanel showCloseButton/>
+  return <ReactQueryDevtoolsPanel props.showCloseButton/>
```

i guess this should be

```diff
-  return <ReactQueryDevtoolsPanel showCloseButton/>
+  return <ReactQueryDevtoolsPanel showCloseButton={props.showCloseButton}/>
```

<details>

<summary>
ManipulationError
</summary>

```
ManipulationError: Manipulation error: A syntax error was inserted.

file.tsx:16:40 - error TS1003: Identifier expected.

16   return <ReactQueryDevtoolsPanel props.showCloseButton/>
                                          ~

Error replacing tree: The children of the old and new trees were expected to have the same count (1:3).
```

</details>

**Comment by juanrgm** [comment]
Thanks you for the contribution to the project!

Some issues to resolve:

1. The current mainstream branch is `develop`.
2. `jest` was replaced by `vitest` in the develop branch.
3. There is no `changeset` file.
4. There are a lot of changes in this PR. This can be a problem for the stability of the proyect because I can't to review all and the changelog will be a bunch of a lot of things (fix, feat, refactor).

Again, thanks for the contribution.

**Comment by milahu** [comment]
> 4\. There are a lot of changes in this PR. This can be a problem for the stability of the proyect because I can't to review all and the changelog will be a bunch of a lot of things (fix, feat, refactor).

thats why we have tests : )
old tests are passing, new features are covered by new tests ...

i am aware that this codemod was made to transform react-material-ui
but its also useful to transform other react components

but this is still a draft (acfba1fbb5c1de2b73e4aa12d004ac42656fc5f5 etc), not ready to merge

**Comment by milahu** [comment]
fix imports for node 19

node 19 has removed the `--experimental-specifier-resolution` flag
https://nodejs.org/nl/blog/announcements/v19-release-announce/#custom-esm-resolution-adjustments

so currently `bin.ts` throws

> Error [ERR_MODULE_NOT_FOUND]: Cannot find module 'lib/actions/fixEsm' imported from lib/bin.js

imo, custom loaders are even worse than the experimental flag
and its simpler to add .js file extensions to all imports = 9b904ec7f96f62b620ef66c99b39099eeaf87085

**Comment by juanrgm** [comment]
> > 4. There are a lot of changes in this PR. This can be a problem for the stability of the proyect because I can't to review all and the changelog will be a bunch of a lot of things (fix, feat, refactor).
> 
> thats why we have tests : ) old tests are passing, new features are covered by new tests ...
> 
> i am aware that this codemod was made to transform react-material-ui but its also useful to transform other react components
> 
> but this is still a draft ([acfba1f](https://github.com/swordev/suid/commit/acfba1fbb5c1de2b73e4aa12d004ac42656fc5f5) etc), not ready to merge

True, but stability is not only the testing phase. Create a pull request with a lot of differents commits is not good:

- Bug fixes are locked.
- Semver is not continuous.
- Collaboration with other users/developers is harder: only one PR (channel of communication) but there is a lot of issues, the discussion is very difficult.
- The review by other developers is impossible because there is not a goal, anything goes.
- The maintenance by the author of the PR will be hard when the base branch be updated.
- Git conflicts are increased.
- Posibility of introduce a bad feature coveraged by test is increased.
- Locating the possible new bugs is more diffcult.
- Reversion in case of problems is harder.
- ...

I really appreciate your dedication, and your work looks great, but I need your help to keep the project in order with this pull request.

**Comment by milahu** [comment]
> Create a pull request with a lot of differents commits is not good

so should i create many PRs?
or add one changeset per commit?

semver is overrated, npm is overrated ...
when people find bugs, they fetch the git history and bisect commits

**Comment by juanrgm** [comment]
The commits should be grouped by goals in multiple pull requests.

The bug fixes have priority and they should separated. If you wrote a test, first add the test commit, and later the fix commit. With this way, the reviewer can run the test without the fix.

The tests increment their complexity slowly. Only one complex test is a bad pattern. The complex test is needed, but in last instance.

If a new feat/fix requires structuring a lot of code, first try do a refactor commit and later the feat/fix commit. This will hep a lot of in the review.

The global changes must be debated with a issue and overall never mixed with others changes.
- Should we add support to Node.js 19?
- Should we use test-snapshot? Are required? It is this a anti-pattern?
- Should we always set the `develop` branch as mainstream for the developing?
- Should we add the `patch-package` dependency? pnpm already has a patch system.

These last issues are global changes and could be implemented later, without blocking other issues more importants.

Each PR must include the `changeset` file to increment the npm package version. This is only necessary if the changes affect to the end user of if this changes are used by other internal packages. A PR can't include several `changeset` files.

If a new react transformer or material component is added it is required regenerate the roadmap file (https://github.com/swordev/suid/blob/main/scripts/actions/genRoadmap.ts).

I know that this may seem complex, but it's very easy. Structuring the commits like this, the 80% of your commits would be merged already.

**Comment by milahu** [comment]
... right. i guess this also belongs to contributing.md

closing as "i dont need this any more"

**Pull request closed without merging**